agent/testflinger_agent/job.py

```diff
@@ -27,239 +35,712 @@
+        self.serial_log = params.rundir / f"{self.phase_id}-serial.log"
+        # phase-specific runner (for executing commands in the rundir)
+        self.runner = CommandRunner(
+            cwd=params.rundir, env=params.client.config
+        )
+        # phase result: initially empty, set later by `JobPhase.run`
+        self.result: Optional[JobPhaseResult] = None
+
+    @abstractmethod
+    def go(self) -> bool:
+        """Return True if the phase should run or False if skipping"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def register(self):
+        """Perform all necessary registrations with the phase runner"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def run_core(self) -> JobPhaseResult:
+        """Execute the "core" of the phase"""
+        raise NotImplementedError
+
+    def process_results(self):
+        """Execute any possible actions after the "core" of the phase"""
+        pass
+
+    def run(self):
+        """This is the generic structure of every phase"""
+
+        assert self.go()
+
+        # register the output handlers with the runner
+        self.runner.register_output_handler(LogUpdateHandler(self.output_log))
+        self.runner.register_output_handler(
+            LiveOutputHandler(
+                self.params.client, self.params.job_data["job_id"]
+            )
+        )
+        # perform additional runner registrations
+        self.register()
+
+        # display phase banner in the Testflinger output
+        for line in self.banner(
+            f"Starting testflinger {self.phase_id} phase "
+            f"on {self.params.client.config.get('agent_id')}"
+        ):
+            self.runner.run(f"echo '{line}'")
+
+        # run the "core" of the phase and store the result
+        self.result = self.run_core()
+
+        # perform any post-core actions
+        self.process_results()
+        self.update_results_file()
+
+    def update_results_file(self):
+        """Update the results file with the results of this phase"""
+        with open(self.results_file, "r+") as results:
+            outcome_data = json.load(results)
+            outcome_data[f"{self.phase_id}_status"] = self.result.exit_code
+            try:
+                with open(self.output_log, "r+", encoding="utf-8") as logfile:
+                    set_truncate(logfile)
+                    outcome_data[f"{self.phase_id}_output"] = logfile.read()
+            except FileNotFoundError:
+                pass
+            try:
+                with open(
+                    self.serial_log, "r+", encoding="utf-8", errors="ignore"
+                ) as logfile:
+                    set_truncate(logfile)
+                    outcome_data[f"{self.phase_id}_serial"] = logfile.read()
+            except FileNotFoundError:
+                pass
+            results.seek(0)
+            json.dump(outcome_data, results)
+
+    def banner(self, line):
```

Comment:
I've always kinda hated the way this works, but still not sure how to best fix it. I see the reason for it, but tricky to replace the functionality. In the meantime though, I think we don't need it to be a method of the class since it never refers to self. Maybe we could move it outside the class as a helper for now instead? (ok to keep in the same module if you like though - still related)